drivers/serial/uart_ns16550.c

```diff
@@ -35,9 +35,36 @@
 #include <drivers/uart.h>
 #include <sys/sys_io.h>
 #include <spinlock.h>
+#include <drivers/pcie/pcie.h>
 
 #include "uart_ns16550.h"
 
+/* Workaround: our UART and PCI subsystems aren't 64 bit clean, and
+ * some systems have firmware that maps the default UART above 4G,
+ * killing console output.
+ *
+ * When set, the 64 bit (!) BAR 0 of the PCI device identified below
+ * will be remapped to the address specified.  Make sure the device is
+ * identified correctly and the address does not collide with any
+ * other memory or devices on the system.
+ */
+#define CONFIG_UART_PCI_REMAP 1
+
+#ifdef CONFIG_UART_PCI_REMAP
+#define PCI_UART_REMAP_BDF PCIE_BDF(0, 0x18, 0)
```

Comment:
could we get this from devicetree?

It's weird and I filed a bug on it https://github.com/zephyrproject-rtos/zephyr/issues/26109, but it looks like we can get the bdf and id for PCI-E using DT_REG_ADDR() and DT_REG_SIZE() for the node instance, see `z_x86_soc_add_mmu_regions()`

Reply:
We could, but really this is just a workaround.  If the PCI and UART code wasn't broken we wouldn't care where the firmware put the device.

Reply:
yes, those macros already defined in the device tree and do the right thing, we should have this just in one place... for example

```
reg = <PCIE_BDF(0,0x18,1) PCIE_ID(0x8086,0x5abe)>;
```